source/common/config/xds_mux/grpc_mux_impl.cc

```diff
@@ -0,0 +1,383 @@
+    sub->second->setControlPlaneIdentifier(response_proto.control_plane().identifier());
+    ENVOY_LOG(debug, "Receiving gRPC updates for {} from {}", response_proto.type_url(),
+              sub->second->controlPlaneIdentifier());
+  }
+
+  pausable_ack_queue_.push(sub->second->handleResponse(response_proto));
+  trySendDiscoveryRequests();
+  Memory::Utils::tryShrinkHeap();
+}
+
+template <class S, class F, class RQ, class RS> void GrpcMuxImpl<S, F, RQ, RS>::start() {
+  ENVOY_LOG(debug, "GrpcMuxImpl now trying to establish a stream");
+  establishGrpcStream();
+}
+
+template <class S, class F, class RQ, class RS>
+void GrpcMuxImpl<S, F, RQ, RS>::handleEstablishedStream() {
+  ENVOY_LOG(debug, "GrpcMuxImpl stream successfully established");
+  for (auto& [type_url, subscription_state] : subscriptions_) {
+    subscription_state->markStreamFresh();
+  }
+  setAnyRequestSentYetInCurrentStream(false);
+  maybeUpdateQueueSizeStat(0);
+  pausable_ack_queue_.clear();
+  trySendDiscoveryRequests();
+}
+
+template <class S, class F, class RQ, class RS>
+void GrpcMuxImpl<S, F, RQ, RS>::handleStreamEstablishmentFailure() {
+  ENVOY_LOG(debug, "GrpcMuxImpl stream failed to establish");
+  // If this happens while Envoy is still initializing, the onConfigUpdateFailed() we ultimately
+  // call on CDS will cause LDS to start up, which adds to subscriptions_ here. So, to avoid a
+  // crash, the iteration needs to dance around a little: collect pointers to all
+  // SubscriptionStates, call on all those pointers we haven't yet called on, repeat if there are
+  // now more SubscriptionStates.
+  absl::flat_hash_map<std::string, S*> all_subscribed;
+  absl::flat_hash_map<std::string, S*> already_called;
+  do {
+    for (auto& [type_url, subscription_state] : subscriptions_) {
+      all_subscribed[type_url] = subscription_state.get();
+    }
+    for (auto& sub : all_subscribed) {
+      if (already_called.insert(sub).second) { // insert succeeded ==> not already called
+        sub.second->handleEstablishmentFailure();
+      }
+    }
+  } while (all_subscribed.size() != subscriptions_.size());
+}
+
+template <class S, class F, class RQ, class RS>
+S& GrpcMuxImpl<S, F, RQ, RS>::subscriptionStateFor(const std::string& type_url) {
+  auto sub = subscriptions_.find(type_url);
+  RELEASE_ASSERT(sub != subscriptions_.end(),
+                 fmt::format("Tried to look up SubscriptionState for non-existent subscription {}.",
+                             type_url));
+  return *sub->second;
+}
+
+template <class S, class F, class RQ, class RS>
+WatchMap& GrpcMuxImpl<S, F, RQ, RS>::watchMapFor(const std::string& type_url) {
+  auto watch_map = watch_maps_.find(type_url);
+  RELEASE_ASSERT(
+      watch_map != watch_maps_.end(),
+      fmt::format("Tried to look up WatchMap for non-existent subscription {}.", type_url));
+  return *watch_map->second;
+}
+
+template <class S, class F, class RQ, class RS>
+void GrpcMuxImpl<S, F, RQ, RS>::trySendDiscoveryRequests() {
```

Comment:
Is it fair to say that this new unified mux is most similar to the existing `NewGrpcMuxImpl`? If a reviewer were to compare the two side-by-side, what would be the key differences or things to pay attention to?

Reply:
> Is it fair to say that this new unified mux is most similar to the existing NewGrpcMuxImpl.

Definitely. Will add some pointers here a bit later.

Reply:
* The most significant source of difference between the implementations is introduction of templates; most of the time there's little to none difference in logic in methods.
* There's no `SubscriptionStuff` in the unified mux implementation; b/c of that creating/accessing subscription state is a bit different.
* Both delta and sotw muxes in the new implementation respect `skip_subsequent_node` flag; the flag is used in`sendGrpcMessage()` method.
* Concrete mux classes are tiny, pretty much all of the logic is in the base class.

Reply:
Thanks, very helpful, CC @adisuissa